### ISSUE TITLE
Base.ReshapedArray ctor segfault with invalid reshape

### ISSUE THREAD

**Issue by oschulz** [JuliaLang/julia#32923]
@biona001 reported this on JuliaArrays/ElasticArrays#14, but it actually concerns  `Base`.

This code will segfault (tested on Linux x86_64, Julia v1.3-alpha):

```julia
M = rand(100, 100);
Mvec = vec(M);
Mnew = Base.ReshapedArray(Mvec, (1000, 1000), ())
```

The reshape is obviously invalid, and I guess users should use `reshape()` instead of calling the ctor directly. Still, it probably shouldn't segfault?

**Comment by oschulz** [comment]
Ah, it's actually `show(Mnew)` that segfaults, not the ctor.

**Comment by mbauman** [comment]
I think this falls into the "don't do that" category.  The same is true of `SubArray` — the constructor is unsafe and error handling happens within `view`.  E.g., displaying `SubArray([], (1:10^18,))` will also segfault.

The advantage of doing it this way is that the constructors are the internal and unsafe high-performance methods (with no error checking), while the friendly functions are the safe interfaces.

I say we just document this.

**Comment by KristofferC** [comment]
> I say we just document this.

I say we just close this. There isn't really any good place to document this imo. `ReshapedArray` is intentionally undocumented and mentioning not to call an undocumented constructor in e.g. `reshape` feels weird.

**Comment by mbauman** [comment]
No, I wouldn't document it in `reshape`.  I have been meaning to add a docstring to `SubArray` — that one is exported — and we could add an internal docstring to `ReshapedArray`, too.

**Comment by KristofferC** [comment]
Adding docstrings to internal types that say "do not call this" just seems redundant. That is already what not having a docstring mean.

**Comment by mbauman** [comment]
Are you trying to bait me into documenting this? Of course it wouldn't be a docstring of simply `"""don't call this"""`, rather it'd have internal documentation about it taking the tuple of dimensions and, optionally, a tuple of multiplicative inverses and that's unsafe as they must be constructed such that subsequent accesses are all in-bounds.  I think that's a great use of an unexported docstring.

**Comment by biona001** [comment]
My original question was what is the best way to reshape a preallocated matrix without reallocating the entire thing. [ElasticArrays](https://github.com/JuliaArrays/ElasticArrays.jl) achieves this for the last dimension, but I need to reshape a square matrix inside a for loop for linear algebra.

**Comment by mbauman** [comment]
Yes, `ReshapedArray` is also very tempting to call as it can be more efficient than `Array`'s builtin reshape in some cases.

**Comment by oschulz** [comment]
> I think this falls into the "don't do that" category.
> I say we just close this.

No argument from me.

One suggestion though: Maybe the internal ctor could take a `Val{:unsafe}` dummy argument, to discourage users from calling it directly?

**Comment by adienes** [comment]
maybe mildly interesting, maybe irrelevant, but this doesn't segfault on Apple M1 (but still does on nightly for intel + linux)

**Comment by mbauman** [comment]
Yeah, not relevant.  It just depends on if it happens to reach outside of the program's memory.  You can probably make it always segfault with larger values.
